src/control/functor/multivariate.lean

```diff
@@ -0,0 +1,770 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad, Mario Carneiro, Simon Hudon
+-/
+import data.fin2
+import logic.function.basic
+import tactic.basic
+
+/-!
+
+Tuples of types, and their categorical structure.
+
+Features:
+
+`typevec n`   : n-tuples of types
+`α ⟹ β`      : n-tuples of maps
+`f ⊚ g`       : composition
+`mvfunctor n` : the type class of multivariate functors
+`f <$$> x`    : notation for map
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type `β` to n-tuple `α` to obtain an (n+1)-tuple
+`drop α`         : drops the last element of an (n+1)-tuple
+`last α`         : returns the last element of an (n+1)-tuple
+`append_fun f g` : appends a function g to an n-tuple of functions
+`drop_fun f`     : drops the last function from an n+1-tuple
+`last_fun f`     : returns the last function of a tuple.
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+
+universes u v w
+
+
+/--
+n-tuples of sorts, as a category
+-/
+def ptypevec (n : ℕ) := fin2 n → Sort*
+
+/-- like `ptypevec` but in `Type*` -/
+def typevec (n : ℕ) := ptypevec.{u+1} n
```

Comment:
I feel like `typevec` reimplements a lot of stuff we already have for `fin`.  We already use `Π i : fin n, α i` to represent heterogeneous tuples, e.g. in vectors and matrices.  We have `fin.cons`, `fin.snoc`, nice notation, etc., for that.

I can't tell if we'd run into issue using the normal `fin` tuples here.  I don't think we absolutely need to refactor this right now unless it's easy, but it's something to keep in mind for later.

Reply:
We did start off with `fin` and it got pretty awkward and hairy and things got simply when we adopted `fin2`. I don't know either how the APIs could be unified. It might be worth seeing if the use of `fin2` expands.